### ISSUE TITLE
warning: Object#untaint is deprecated and will be removed in Ruby 3.2

### ISSUE THREAD

**Issue by redconfetti** [tzinfo/tzinfo#145]
The following warning just came up many many times after I upgraded to jruby v9.4.0.0 ([Ruby 3.1|https://www.jruby.org/2022/11/23/jruby-9-4-0-0.html])
```
/Users/johndoe/Developer/my_app/vendor/bundle/jruby/3.1.0/gems/tzinfo-2.0.5/lib/tzinfo/data_sources/zoneinfo_data_source.rb:447: warning: Object#untaint is deprecated and will be removed in Ruby 3.2
```

This warning is related to a call in [TzInfo::DataSource::ZoneinfoDataSource#enum_timezones](https://github.com/tzinfo/tzinfo/blob/v2.0.5/lib/tzinfo/data_sources/zoneinfo_data_source.rb#L447)

**Comment by headius** [comment]
I have pushed a fix for JRuby 9.4.1.0 that allows this peculiar refinement call to work. The problem was that we only detect refined scopes when there's a direct `using` call in a parent scope. The `send(:using, ...)` was not detected as such, so we never made refined calls. My fix in jruby/jruby#7599 treats a `send` of `:using` or `:refine` as though they are direct calls, and activates refinements in the current scope.

With this, tzinfo tests run without the untaint warnings:

```
$ bundle exec rake test
WARNING: Private key not found. Not signing gem file.
/Users/headius/work/jruby/lib/ruby/gems/shared/gems/concurrent-ruby-1.2.0/lib/concurrent-ruby/concurrent/executor/java_thread_pool_executor.rb:14: warning: method redefined; discarding old to_int
/Users/headius/work/jruby/lib/ruby/gems/shared/gems/concurrent-ruby-1.2.0/lib/concurrent-ruby/concurrent/executor/java_thread_pool_executor.rb:14: warning: method redefined; discarding old to_f
Using Ruby DataSource: tzdb v2020d, tzinfo-data v1.2020.4.test
Run options: --seed 17735

# Running:

........................................................................................................................................................................................................................................................................................................................................................................S..........................................................S.....................................................................................................S..................................................................................S................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................S..S..........................S.....S.............................................................S......S..........................S....................................................S.........................................................................................................................................S................................................................................................................................................................................................................

Finished in 33.959832s, 53.7400 runs/s, 390.4613 assertions/s.

1825 runs, 13260 assertions, 0 failures, 0 errors, 13 skips

You have skipped tests. Run with --verbose for details.
/Users/headius/work/jruby/lib/ruby/gems/shared/gems/concurrent-ruby-1.2.0/lib/concurrent-ruby/concurrent/executor/java_thread_pool_executor.rb:14: warning: method redefined; discarding old to_int
/Users/headius/work/jruby/lib/ruby/gems/shared/gems/concurrent-ruby-1.2.0/lib/concurrent-ruby/concurrent/executor/java_thread_pool_executor.rb:14: warning: method redefined; discarding old to_f
Using Ruby DataSource: tzdb v2020d, tzinfo-data v2.2020.4.test
Run options: --seed 48391

# Running:

.......................................................................S..........................................................................................................................................................................................................................................................................................S...........S............................................................................................................................................S.......................S...................................................................................................................................................S...S..................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................S...........................................S.............................S...............................................S...................................................S................................................................................................................................................................................................................S.................................................................................................................................................................................................................................................................

Finished in 27.477894s, 66.3442 runs/s, 482.4242 assertions/s.

1823 runs, 13256 assertions, 0 failures, 0 errors, 13 skips

You have skipped tests. Run with --verbose for details.
/Users/headius/work/jruby/lib/ruby/gems/shared/gems/concurrent-ruby-1.2.0/lib/concurrent-ruby/concurrent/executor/java_thread_pool_executor.rb:14: warning: method redefined; discarding old to_int
/Users/headius/work/jruby/lib/ruby/gems/shared/gems/concurrent-ruby-1.2.0/lib/concurrent-ruby/concurrent/executor/java_thread_pool_executor.rb:14: warning: method redefined; discarding old to_f
Using Zoneinfo DataSource: /Users/headius/work/tzinfo/test/zoneinfo
Run options: --seed 5875

# Running:

.........................................................................................................................................................................................................................................................................................................................................................................................................SS...........................................................................................................................................................................................................................................S...................................................................................................................................................................................................S......................................................................S........................S....................................................................................................................................................................................................................................................................................................................................................S.......S..........................................................................................................SS.............................................................................................................................................................................................................................................................................................................S.......S....S................................................................................................................................

Finished in 27.400007s, 66.5328 runs/s, 483.8685 assertions/s.

1823 runs, 13258 assertions, 0 failures, 0 errors, 13 skips

You have skipped tests. Run with --verbose for details.
```

**Comment by philr** [comment]
Thanks for the report @redconfetti and thanks for the fix to JRuby @headius.

As @headius alluded to, tzinfo was using [a refinement](https://github.com/tzinfo/tzinfo/blob/v2.0.5/lib/tzinfo/untaint_ext.rb) to define or redefine `Object#using` when it is missing or a no-op method that raises a warning. The refinement was being used with [`send(:using, UntaintExt)`](https://github.com/tzinfo/tzinfo/blob/v2.0.5/lib/tzinfo/data_sources/zoneinfo_data_source.rb#L7) as a workaround for issues with old versions of JRuby. JRuby v9.4.0.0 wasn't detecting this peculiar arrangement as using the refinement, so was still calling the underlying `Object#using` when it should have been using the redefinition.

I've released [v2.0.6](https://github.com/tzinfo/tzinfo/releases/tag/v2.0.6) (and [v1.2.11](https://github.com/tzinfo/tzinfo/releases/tag/v1.2.11)) with a change to replace the refinement (and its peculiar way of being used) with a call to an internal [`RubyCoreSupport.untaint`](https://github.com/tzinfo/tzinfo/blob/v2.0.6/lib/tzinfo/ruby_core_support.rb#L10) method that calls `Object#untaint` when necessary. This eliminates the warnings with JRuby v9.4.0.0. I think this is a simpler approach and retains support for old JRuby versions.

**Comment by headius** [comment]
Thank you for the summary, it matches up with what I understood after looking at the code. The fix in JRuby is appropriate and we're fine shipping that but I also appreciate your working around this unusual issue on your end. Looks like we're good to go!
